codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java

```diff
@@ -0,0 +1,265 @@
+        verifyNoMoreInteractions(frameWriter);
+        channel.writeInbound(Http2CodecUtil.connectionPrefaceBuf());
+        frameListener.onSettingsRead(http2HandlerCtx, new Http2Settings());
+        verify(frameWriter).writeSettingsAck(eq(http2HandlerCtx), anyChannelPromise());
+        frameListener.onSettingsAckRead(http2HandlerCtx);
+    }
+
+    @Test
+    public void startStop() throws Exception {
+        assertTrue(channel.isActive());
+        channel.close();
+        verify(frameWriter).writeGoAway(
+                eq(http2HandlerCtx), eq(0), eq(0L), eq(Unpooled.EMPTY_BUFFER), anyChannelPromise());
+        assertTrue(!channel.isActive());
+    }
+
+    @Test
+    public void headerRequestHeaderResponse() throws Exception {
+        LastInboundHandler stream = new LastInboundHandler();
+        streamInit.handler = stream;
+        frameListener.onHeadersRead(http2HandlerCtx, 1, request, 31, true);
+        assertNull(streamInit.handler);
+        assertEquals(new DefaultHttp2HeadersFrame(request, true, 31), stream.readInbound());
+        assertNull(stream.readInbound());
+        assertTrue(stream.channel().isActive());
+
+        stream.writeOutbound(new DefaultHttp2HeadersFrame(response, true, 27));
+        verify(frameWriter).writeHeaders(
+                eq(http2HandlerCtx), eq(1), eq(response), anyInt(), anyShort(), anyBoolean(),
+                eq(27), eq(true), anyChannelPromise());
+        verify(frameWriter, never()).writeRstStream(
+            any(ChannelHandlerContext.class), anyInt(), anyLong(), anyChannelPromise());
+        assertFalse(stream.channel().isActive());
+        assertTrue(channel.isActive());
+    }
+
+    @Test
+    public void entityRequestEntityResponse() throws Exception {
+        LastInboundHandler stream = new LastInboundHandler();
+        streamInit.handler = stream;
+        frameListener.onHeadersRead(http2HandlerCtx, 1, request, 0, false);
+        assertEquals(new DefaultHttp2HeadersFrame(request, false), stream.readInbound());
+        assertNull(stream.readInbound());
+        assertTrue(stream.channel().isActive());
+
+        ByteBuf hello = bb("hello");
+        frameListener.onDataRead(http2HandlerCtx, 1, hello, 31, true);
+        // Release hello to emulate ByteToMessageDecoder
+        hello.release();
+        Http2DataFrame inboundData = stream.readInbound();
+        assertEquals(new DefaultHttp2DataFrame(bb("hello"), true, 31), inboundData);
+        assertEquals(1, inboundData.refCnt());
+        assertNull(stream.readInbound());
+        assertTrue(stream.channel().isActive());
+
+        stream.writeOutbound(new DefaultHttp2HeadersFrame(response, false));
+        verify(frameWriter).writeHeaders(eq(http2HandlerCtx), eq(1), eq(response), anyInt(),
+            anyShort(), anyBoolean(), eq(0), eq(false), anyChannelPromise());
+        assertTrue(stream.channel().isActive());
+
+        stream.writeOutbound(new DefaultHttp2DataFrame(bb("world"), true, 27));
```

Comment:
@Scottmitch, but `dataFrame.content()` _won't_ throw because `retain()` was called on `bb("world")`. Since they share the same `refCnt()` the `assertEquals(1, outboundData.getValue().refCnt());` later is checking `dataFrame.refCnt()` at the same time.

I guess I could assert `dataFrame.refCnt() == 0` _after_ `outboundData.getValue().release();`, but that seems pretty hard to follow.